data-model-solr/src/main/java/org/mousephenotype/cda/solr/service/PostQcService.java

```diff
@@ -121,6 +128,141 @@ public void postSetup() {
+     * @return
+     */
+    public JSONObject getPleiotropyMatrix(List<String> topLevelMpTerms) throws IOException, SolrServerException {
+
+        TreeMap<String, TreeMap<String, Integer>> matrix = new TreeMap<>();
+        TreeMap<String, Set<String>> genesByTopLevelMp = new TreeMap<>(); // <mp, <genes>>
+        SolrQuery query = getPleiotropyQuery(topLevelMpTerms);
+
+        try {
+
+            QueryResponse queryResponse = solr.query(query, SolrRequest.METHOD.POST);
+            Set<String> facets = getFacets(queryResponse).get(GenotypePhenotypeDTO.TOP_LEVEL_MP_TERM_NAME).keySet();
+
+            // Fill matrix with 0s
+            for (String facet : facets){
+                TreeMap<String, Integer> row = new TreeMap<>();
+                for (String f : facets) {
+                    row.put(f, 0);
+                }
+                matrix.put(facet, row);
+                genesByTopLevelMp.put(facet, new HashSet<>());
+            }
+
+            Map<String, List<String>> facetPivotResults = getFacetPivotResults(queryResponse, query.get("facet.pivot"));
+            for (String gene : facetPivotResults.keySet()) {
+                List<String> mpTerms = facetPivotResults.get(gene);
+
+                for (String mpA : mpTerms) {
+                    Set<String> temp = genesByTopLevelMp.containsKey(mpA) ? genesByTopLevelMp.get(mpA) : new HashSet<>();
+                    temp.add(gene);
+                    genesByTopLevelMp.put(mpA, temp);
+
+                    if (mpTerms.size() > 1) { // other phenotypes too
+                        for (String mpB : mpTerms) {
+                            if (!mpA.equalsIgnoreCase(mpB)) {
+                                matrix = add(mpA, mpB, 1, matrix);
+                            }
+                        }
+                    } else { // only one top level mp for this gene. count as self
+                        matrix = add(mpA, mpA, 1, matrix);
+                    }
+                }
+            }
+
+        } catch (SolrServerException | IOException e) {
+            e.printStackTrace();
+        }
+
+        List<JSONObject> labelList = genesByTopLevelMp.entrySet().stream().map(entry -> new JSONObject().put("name", entry.getKey()).put("geneCount", entry.getValue().size())).collect(Collectors.toList());
+
+        JSONArray jsonMatrix = new JSONArray();
+        for (String keyA : matrix.keySet()){
+            JSONArray row = new JSONArray();
+            for (String keyB : matrix.keySet() ){
+                row.put(matrix.get(keyA).get(keyB));
+            }
+            jsonMatrix.put(row);
+        }
+
+        JSONObject result = new JSONObject();
+        result.put("matrix", jsonMatrix);
+        result.put("labels", labelList);
+
+        return result;
+
+    }
+
+    public String getPleiotropyDownload(List<String> topLevelMpTerms) throws IOException, SolrServerException {
+
+        TreeMap<String, TreeMap<String, Integer>> matrix = new TreeMap<>();
```

Comment:
I don't know about "needed" but I changed it to a matrix. Still think it's harder to follow a bidimensional array and it's parallel label array than just a map.